docs/keycloak.md

```diff
@@ -0,0 +1,356 @@
+                    if(rep.getClientRoles().get(clientId)) newGroup.put("clientRoles",rep.getClientRoles().get(clientId));
+                } else {
+                    var clients = realm.getClients();
+                     for (i= 0; i<clients.size(); i++){
+                        item = clients.get(i);
+                        var clientId = item.getClientId();
+                        if(rep.getClientRoles().get(clientId)) newGroup.put(clientId,rep.getClientRoles().get(clientId));
+                    }
+                }
+                groups.add(newGroup);
+            }
+        }
+    }
+    if( group.getParent() !== null ) {
+        scanGroups(group.getParent());
+    }
+}
+
+function getTenant(name){
+    for (i= 0; i<tenants.size(); i++){
+        item = tenants.get(i);
+        if (item.get("name")===name)
+            return item;
+    }
+    return null;
+}
+
+function getGroup(groups, name){
+    for (i= 0; i<groups.size(); i++){
+        item = groups.get(i);
+        if (item.get("name")===name)
+            return item;
+    }
+    return null;
+}
+
+function getRoot(group){
+    if(group.getParent() !== null)
+        return getRoot(group.getParent());
+    return group;
+}
+
+
+function isServicePath(group){
+    if (group.getFirstAttribute("servicePath") == "true" || group.getFirstAttribute("servicepath") == "true")  
+        return true;
+    return false;
+}
+
+function isTenant(group){
+    if (group.getParent()===null &&
+        group.getFirstAttribute("tenant") == "true"){
+        return true;
+    }
+    return false;
+}
+
+token.setOtherClaims("tenants", tenants);
+```
+
+In addition, under the scope **Tenant-names** the script **Only-tenant-names**
+was created:
+
+```javascript
+var HashMap = Java.type('java.util.HashMap');
+var HashSet = Java.type('java.util.HashSet');
+var ArrayList = Java.type('java.util.ArrayList');
+var ModelToRepresentation = Java.type('org.keycloak.models.utils.ModelToRepresentation');
+var groups = user.getGroups();
+var tenants = new ArrayList();
+
+groups.forEach(scanTenant);
+
+function scanTenant(group){
+    if ( isTenant(group)===true ) {
+        found = getTenant(group.getName());
+        if ( !found ) {
+            tenants.add(group.getName());
+        }
+    } else if( group !== null && group.getParent() !== null ) {
+        scanTenant(group.getParent());
+    }
+}
+
+function getTenant(name){
+    for (i= 0; i<tenants.size(); i++){
+        item = tenants.get(i);
+        if (item===name)
+            return item;
+    }
+    return null;
+}
+
+function getRoot(group){
+    if(group.getParent() !== null)
+        return getRoot(group.getParent());
+    return group;
+}
+
+
+function isTenant(group){
+    if (group.getParent()===null &&
+        group.getFirstAttribute("tenant") == "true"){
+        return true;
+    }
+    return false;
+}
+
+token.setOtherClaims("tenants", tenants);
+```
+
+## Authentication Flows
+
```

Comment:
explain that to ensure that for specific services (list which ones), only user with a given role can access, this auth flow was created

Reply:
maybe you can also make a diagram to explain how it works

Reply:
basically the there is an oauth_proxy (check the repository) that allows you to authenticate against an OIDC provider for any http end point. this is needed when web ui don't support oidc natively. moreover the role check is needed because this UI not supporting OIDC do not know what to do with OIDC returned role, so without that anyone with an oidc account will just have access to the service.

Reply:
so let's have a main section here called:

protecting Web applications that don't support natively OIDC/AUTH

first explain the general approach (i mentioned in the prev comment)
secondly provide and example of auth proxy with config.
thirdly describe the configuration (in a sub section)

Reply:
I'm still not sure what this part is meant to be, I haven't done anything on that front.

Reply:
k, leave it :) i will doc, can you just list though the clients using the custom auth flow?